src/lib/transition_frontier/transition_frontier0.ml

```diff
@@ -0,0 +1,689 @@
+                  ~expect:(Inputs.Ledger_proof.statement proof_data).source
+                  ( Ledger.Db.merkle_root t.root_snarked_ledger
+                  |> Frozen_ledger_hash.of_ledger_hash ) ;
+                let db_mask = Ledger.of_database t.root_snarked_ledger in
+                Non_empty_list.iter txns ~f:(fun txn ->
+                    (* TODO: @cmr use the ignore-hash ledger here as well *)
+                    Ledger.apply_transaction db_mask txn
+                    |> Or_error.ok_exn |> ignore ) ;
+                (* TODO: See issue #1606 to make this faster *)
+                Ledger.commit db_mask ;
+                ignore
+                  (Ledger.Maskable.unregister_mask_exn
+                     (Ledger.Any_ledger.cast
+                        (module Ledger.Db)
+                        t.root_snarked_ledger)
+                     db_mask)
+            | _, false | None, _ -> () ) ;
+            [%test_result: Frozen_ledger_hash.t]
+              ~message:
+                "Root snarked ledger hash diverged from blockchain state \
+                 after root transition"
+              ~expect:
+                (Consensus.Blockchain_state.snarked_ledger_hash
+                   (Breadcrumb.blockchain_state new_root_node.breadcrumb))
+              ( Ledger.Db.merkle_root t.root_snarked_ledger
+              |> Frozen_ledger_hash.of_ledger_hash ) ;
+            Some (garbage, new_root_node) )
+          else None
+        in
+        let maybe_new_best_tip_bc =
+          if node.length > best_tip_node.length then
+            Some best_tip_node.breadcrumb
+          else None
+        in
+        match (maybe_new_root, maybe_new_best_tip_bc) with
+        | Some (garbage, new_root_node), best_tip ->
+            Transition_frontier_diff.New_root
+              { old_root= root_node.breadcrumb
+              ; new_root= new_root_node.breadcrumb
+              ; added= node.breadcrumb
+              ; garbage=
+                  List.map garbage ~f:(fun g ->
+                      (Hashtbl.find_exn t.table g).breadcrumb )
+              ; old_best_tip= best_tip }
+        | _, Some new_best_tip_bc ->
+            Transition_frontier_diff.New_best_tip
+              { new_best_tip= new_best_tip_bc
+              ; old_best_tip= best_tip_node.breadcrumb }
+        | None, _ -> Transition_frontier_diff.Extend_best_tip node.breadcrumb
+    )
+
+  let clear_paths t = Hashtbl.clear t.table
+
+  let best_tip_path_length_exn {table; root; best_tip; _} =
+    let open Option.Let_syntax in
+    let result =
+      let%bind best_tip_node = Hashtbl.find table best_tip in
+      let%map root_node = Hashtbl.find table root in
+      best_tip_node.length - root_node.length
+    in
+    result |> Option.value_exn
+
+  module For_tests = struct
+    let root_snarked_ledger {root_snarked_ledger; _} = root_snarked_ledger
+  end
+end
+
+let%test_module "Transition_frontier tests" =
```

Comment:
You can basically use this. I don't think we would ever need to use this unit test anymore.